### PR TITLE
Added webcam support for Linux

### PR DESCRIPTION
**Pull request by Schmetzler**
I added webcam support for Linux.
I used libv4l2 which is dynamically loaded if present. If it is not present the kernel default v4l2 library is used.
It should be able to grab images even without libv4l2 if the camera supports RGB output directly.
With libv4l2 the image is always grabbed as RGB8 image.

I also implemented a simple hotplug feature which just checks the /dev/video* files every second.

I tested it with the TestCameraServer Project from [https://github.com/BastiaanOlij/godot3_test_projects.git](https://github.com/BastiaanOlij/godot3_test_projects.git)
I could only test for TYPE_IO_MMAP as my camera only supports this, but I implemented also the other types using the information provided by [https://kernel.readthedocs.io/en/latest/media/uapi/v4l/capture.c.html](https://kernel.readthedocs.io/en/latest/media/uapi/v4l/capture.c.html)

- *Production edit: This partially addresses https://github.com/godotengine/godot/issues/46531.*

### CODE REVIEW

File: modules/camera/camera_x11.cpp [hunk 1]
```diff
@@ -0,0 +1,770 @@
...
+	device = NULL;
+};
+
+void CameraFeedX11::set_device(V4l2_Device *p_device) {
+	device = p_device;
+
+	// get some info
+	name = device->name;
+	// I do not think that another position is possible on linux
+	position = CameraFeed::FEED_UNSPECIFIED;
```
Comment (BastiaanOlij) [inline modules/camera/camera_x11.cpp:602]: Unspecified I think is fine for a webcam, front/back really only applies to phones.

**Review by BastiaanOlij — APPROVED** [review]
I'm not a linux guy so I can't really test whether this works as advertised but doing a quick code review it looked sensible. 

Would like someone to look at this who knows more about linux, @hpvb maybe you can have a quick look?

File: modules/camera/camera_x11.cpp [hunk 2]
```diff
@@ -0,0 +1,770 @@
...
+std::vector<__u32> supported_formats{
+	V4L2_PIX_FMT_RGB24
+};
+
+int V4l2_Device::xioctl(int fd, unsigned long int request, void *arg) {
+	int r;
+	do {
+		r = funcs->ioctl(fd, request, arg);
+	} while (-1 == r && (EINTR == errno || errno == EAGAIN));
+	return r;
```
Comment (BastiaanOlij) [inline modules/camera/camera_x11.cpp:67]: yoda coding! love it but not really our style :)

File: modules/camera/camera_x11.cpp [hunk 3]
```diff
@@ -0,0 +1,770 @@
...
+	// Find available cameras we have at this time
+	update_feeds();
+
+	// start the hotplug thread
+	// alive is also used to trigger debug output
+	// (as devices that are incompatible would always be checked
+	// and therefore print always debug messages)
+	alive = true;
+	hotplug_thread = std::thread(&CameraX11::check_change, this);
+};
```
Comment (BastiaanOlij) [inline modules/camera/camera_x11.cpp:757]: maybe only start the thread if library was loaded successfully?

File: modules/camera/camera_x11.cpp (hunk 2 above)
Comment (Schmetzler) [reply modules/camera/camera_x11.cpp:67]: TBH that was just how the example was implemented. I found a little weird too, but hey semantics are the same.^^ I can change this and make it more common if needed.

File: modules/camera/camera_x11.cpp (hunk 3 above)
Comment (Schmetzler) [reply modules/camera/camera_x11.cpp:757]: Well I use both, the libv4l2 library (which is an LGPL wrapper above v4l2) and if it is not present, I use v4l2 (which is in the linux kernel). It should be possible to read from a camera using only v4l2 if it provides the images already in RGB format.
So there is a change the webcam would work even if libv4l2 is not installed.

**Comment by Schmetzler** [comment]
Ah as I want to mention. I would also like to push it to the godot 3.x branch, but therefore the Vector<uint8_t> must be changed to PoolVector<uint8_t>. But i guess it should be cherrypicked after this one is merged?

File: modules/camera/camera_x11.cpp (hunk 2 above)
Comment (BastiaanOlij) [reply modules/camera/camera_x11.cpp:67]: I'll let @akien-mga be the judge of coding styles :)

I do think Yoda coding has merit, it does catch certain errors, but I don't think we've got it anywhere else in the source code.

File: modules/camera/camera_x11.cpp (hunk 3 above)
Comment (BastiaanOlij) [reply modules/camera/camera_x11.cpp:757]: ah ok cool :)

**Comment by BastiaanOlij** [comment]
> 
> 
> Ah as I want to mention. I would also like to push it to the godot 3.x branch, but therefore the Vector<uint8_t> must be changed to PoolVector<uint8_t>. But i guess it should be cherrypicked after this one is merged?

Yes that would be a cherry pick :)

File: modules/camera/camera_x11.cpp (hunk 2 above)
Comment (Calinou) [reply modules/camera/camera_x11.cpp:67]: Indeed, we don't use yoda conditions in Godot's codebase as far as I know.

Comment (Schmetzler) [reply modules/camera/camera_x11.cpp:67]: So I removed the Yoda conditions to not confuse the codebase.

**Comment by Ansraer** [comment]
Just gonna link the webcam support issue here. Should make it easier for me to adjust the issue when this gets merged: #46531

**Comment by georgewsinger** [comment]
Does this code support setting the scene's `Environment` to be the webcam stream?  I have tried to do this and am just getting a black sky.

**Code source:** I merged this code into our Godot 3.2 fork here:

- [Commit adding this code](https://github.com/SimulaVR/godot/commit/83460bd0c08778ea6213eaf96997d18577141208) (doesn't compile in Godot 3.2)
- [Commit adjusting Vector to PoolVector to get to work on Godot 3.2](https://github.com/SimulaVR/godot/commit/da6a6e7ec0eced301170665a950e4b94518a687f) (does compile in Godot 3.2)

Using this to run godot with a webcam located at `/dev/video0` (verified to work outside of godot) and with

```
environment.set_background(BG_CAMERA_FEED)
```

set results in

```
libv4l2 found.
Registered camera UVC Camera (046d:0825) (/dev/video0) with id 1 position 0 at index 0
/dev/video1 is no video capture device.
```

but with a **black background**; moreover, the webcam doesn't even turn on.

**Tracing the source of the issue:** I've discovered that the specific failure is located in [RasterizerSceneGLES3::render_scene](https://github.com/SimulaVR/godot/blob/76dbaca3ceb366411ed36d7df5f24b9aec113f29/drivers/gles3/rasterizer_scene_gles3.cpp#L4459), where `feed.is_valid()` evaluates to `false`, causing the webcam texture to not get drawn in the background/`Environment`:

```C++
				if (feed.is_valid() && (feed->get_base_width() > 0) && (feed->get_base_height() > 0)) {
					// copy our camera feed to our background

					glDisable(GL_BLEND);
					glDepthMask(GL_FALSE);
					glDisable(GL_DEPTH_TEST);
					glDisable(GL_CULL_FACE);
// ...
```

**Comment by georgewsinger** [comment]
After some fidgeting I got this to work, but only from gdb: https://www.youtube.com/watch?v=KhTs9uriXDQ

```C++
bool CameraFeedX11::activate_feed() {
	// activate streaming if not already
	if (!device->streaming) {
		if (!device->check_device()) {
			return false;
		}
		if (!device->request_buffers()) {
			device->close();
			return false;
		}
		if (!device->start_streaming(this)) {
			device->cleanup_buffers();
			device->close();
			return false;
		}
	};

	return true;
};
```

If I manually step through this via gdb (`n` over and over through this function), then the webcam starts.  If I don't use gdb to step through like this, then it fails.  There must be some sort of async issue going where pauses are necessary to get it to work.

**Comment by Calinou** [comment]
> If I manually step through this via gdb (n over and over through this function), then the webcam starts. If I don't use gdb to step through like this, then it fails. There must be some sort of async issue going where pauses are necessary to get it to work.

Does it work if you call `OS::get_singleton()->delay_msec(100)` every now and then in that function?

**Comment by georgewsinger** [comment]
> > If I manually step through this via gdb (`n` over and over through this function), then the webcam starts. If I don't use gdb to step through like this, then it fails. There must be some sort of async issue going where pauses are necessary to get it to work.
> 
> Does it work if you call `OS::get_singleton()->delay_msec(100)` every now and then in that function (with possibly larger or smaller values)?

Yes, changing the function to the following makes now it work without having to manually step through it w/GDB:

```C++
bool CameraFeedX11::activate_feed() {
	// activate streaming if not already
	if (!device->streaming) {
		OS::get_singleton()->delay_usec(100);
		if (!device->check_device()) {
			return false;
		}
		OS::get_singleton()->delay_usec(100);
		if (!device->request_buffers()) {
			device->close();
			return false;
		}
		OS::get_singleton()->delay_usec(100);
		if (!device->start_streaming(this)) {
			device->cleanup_buffers();
			device->close();
			return false;
		}
	};

	return true;
};
```

**Comment by georgewsinger** [comment]
```C++
	fd = funcs->open(dev_name.c_str(), O_RDWR | O_NONBLOCK, 0);
```

`O_NONBLOCK` is used in the calls to `funcs->open`, which is probably what's causing the async issue here.

**Comment by georgewsinger** [comment]
The webcam image is also left-right inverted when shown in the Environment.

**Comment by RMKD** [comment]
Hey all - excited to see this. I'm trying to reproduce with a build from the fork. I get both a usb web cam and CV1 external cameras detected with libv4l2 in the log and can get them via `CameraServer.feeds()` but can't figure out how to activate or retrieve the texture in gdscript. Do I need to do something in native?

**Comment by MrJustreborn** [comment]
Will this work with a v4l2loopback-device? (eg Screencapture/OBS)

**Comment by Schmetzler** [comment]
@georgewsinger 
> The webcam image is also left-right inverted when shown in the Environment, which is fixed by a call to `img->flip_x()` in `V4l2_Device::get_image`.

I don't think this should be done in the module, because sometimes it depends on the webcam what kind of picture it sends. In my opinion such things should be done in the shader.

> ```c++
> 	fd = funcs->open(dev_name.c_str(), O_RDWR | O_NONBLOCK, 0);
> ```
> 
> `O_NONBLOCK` is used in the calls to `funcs->open`, which is probably what's causing the async issue here.

To solve this async issue I guess your proposal with the little delay is best. What do the others think. I do not think that making the open call blocking is a good idea

@MrJustreborn 
> Will this work with a v4l2loopback-device? (eg Screencapture/OBS)

Yes it works also with a v4l2loopback device (I tested this)

**Comment by RMKD** [comment]
I'm still missing something in terms of how to get a basic test scene running. It seems like CameraServer should be returning feeds as CameraX11 objects, but invoking functions on the returned object yields a 'non-existent function' error. Any suggestions for setup or scene links?

Also, should `instance` be `instantiate` with the current head of master as the compiler suggests? Possibly there is a breaking change made after this PR was authored?

```
[ 65%] modules/camera/camera_x11.cpp: In member function 'void V4l2_Device::get_image(Ref<CameraFeed>, uint8_t*)':
modules/camera/camera_x11.cpp:556:13: error: 'class Ref<Image>' has no member named 'instance'; did you mean 'instantiate'?
  556 |         img.instance();
      |             ^~~~~~~~
      |             instantiate
[ 65%] Compiling ==> editor/rename_dialog.cpp
modules/camera/camera_x11.cpp: In member function 'void CameraX11::update_feeds()':
modules/camera/camera_x11.cpp:690:41: error: 'class Ref<CameraFeedX11>' has no member named 'instance'; did you mean 'instantiate'?
  690 |                                 newfeed.instance();
      |                                         ^~~~~~~~
      |                                         instantiate
scons: *** [modules/camera/camera_x11.linuxbsd.tools.64.o] Error 1
scons: building terminated because of errors.
```

**Comment by Calinou** [comment]
> Also, should instance be instantiate with the current head of master as the compiler suggests?

Indeed, `instance()` was renamed to `instantiate()` in `master`.

**Comment by RMKD** [comment]
Updating the `.cpp` to use `instiantiate()` compiles just fine (thx for the detailed change list).  If i include `GDCLASS(CameraFeedX11, CameraFeed);` in the header file the server and feeds are recognized as the X11 class types... but they still don't have the expected methods from `CameraFeed` like `get_name()` - I notice some of the other modules include code to explicitly bind methods along with the `GDCLASS` clause:
```
//.h
protected:
	static void _bind_methods();
```
```
//.cpp
void GLTFTexture::_bind_methods() {
	ClassDB::bind_method(D_METHOD("get_src_image"), &GLTFTexture::get_src_image);
	ClassDB::bind_method(D_METHOD("set_src_image", "src_image"), &GLTFTexture::set_src_image);
```
I'm a little out of my depth on the c++ front so making some guesses - is something like the above required to invoke inherited class functions from gdscript in v4?

**Comment by RMKD** [comment]
well, this explains a lot: https://github.com/godotengine/godot/blob/master/servers/camera/camera_feed.cpp#L36-L37 (turns out manually binding the methods in `camera_x11.cpp` _does_ help... because the parent class currently disables them in master)

**Comment by RMKD** [comment]
From what I can piece together, the circumvented `CameraFeed` code is impacted by both the refactoring of VisualServer (to RenderingServer and maybe some other places?) and the noted Vulkan refactoring.  The required porting appears to include: 
- `RenderingServer->texture_create()`
- `RenderingServer->texture_allocate()`
- `RenderingServer->texture_set_data()`

Is there a thread of work covering this or would it just be an isolated PR?
Are [rendering_device](https://github.com/godotengine/godot/blob/46beaacec3d18c0d5f1d744e2c63ca09f5152c85/servers/rendering/rendering_device.cpp#L83) or [renderer_storage](https://github.com/godotengine/godot/blob/7bcfc66fb4253fb121008809e4e81c5b704776b7/servers/rendering/renderer_storage.h#L127-L129) the right places to look for replacements?

**Comment by bruvzg** [comment]
A patch for camera feed (tested only on macOS), base size parameters or CameraTexture class might need some additional changes, currently it's not matching half size CbCr texture, but you need a custom shader to combine textures, so it probably not an issue:

```diff
diff --git a/servers/camera/camera_feed.cpp b/servers/camera/camera_feed.cpp
index eab4c61591..0600add6d5 100644
--- a/servers/camera/camera_feed.cpp
+++ b/servers/camera/camera_feed.cpp
@@ -33,8 +33,6 @@
 #include "servers/rendering_server.h"
 
 void CameraFeed::_bind_methods() {
-// FIXME: Disabled during Vulkan refactoring, should be ported.
-#if 0
 	// The setters prefixed with _ are only exposed so we can have feeds through GDNative!
 	// They should not be called by the end user.
 
@@ -54,7 +52,6 @@ void CameraFeed::_bind_methods() {
 
 	ClassDB::bind_method(D_METHOD("_set_RGB_img", "rgb_img"), &CameraFeed::set_RGB_img);
 	ClassDB::bind_method(D_METHOD("_set_YCbCr_img", "ycbcr_img"), &CameraFeed::set_YCbCr_img);
-	ClassDB::bind_method(D_METHOD("_allocate_texture", "width", "height", "format", "texture_type", "data_type"), &CameraFeed::allocate_texture);
 
 	ADD_GROUP("Feed", "feed_");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "feed_is_active"), "set_active", "is_active");
@@ -68,7 +65,6 @@ void CameraFeed::_bind_methods() {
 	BIND_ENUM_CONSTANT(FEED_UNSPECIFIED);
 	BIND_ENUM_CONSTANT(FEED_FRONT);
 	BIND_ENUM_CONSTANT(FEED_BACK);
-#endif
 }
 
 int CameraFeed::get_id() const {
@@ -144,14 +140,6 @@ CameraFeed::CameraFeed() {
 	datatype = CameraFeed::FEED_RGB;
 	position = CameraFeed::FEED_UNSPECIFIED;
 	transform = Transform2D(1.0, 0.0, 0.0, -1.0, 0.0, 1.0);
-
-// FIXME: Disabled during Vulkan refactoring, should be ported.
-#if 0
-	// create a texture object
-	RenderingServer *vs = RenderingServer::get_singleton();
-	texture[CameraServer::FEED_Y_IMAGE] = vs->texture_create(); // also used for RGBA
-	texture[CameraServer::FEED_CBCR_IMAGE] = vs->texture_create();
-#endif
 }
 
 CameraFeed::CameraFeed(String p_name, FeedPosition p_position) {
@@ -164,33 +152,21 @@ CameraFeed::CameraFeed(String p_name, FeedPosition p_position) {
 	datatype = CameraFeed::FEED_NOIMAGE;
 	position = p_position;
 	transform = Transform2D(1.0, 0.0, 0.0, -1.0, 0.0, 1.0);
-
-// FIXME: Disabled during Vulkan refactoring, should be ported.
-#if 0
-	// create a texture object
-	RenderingServer *vs = RenderingServer::get_singleton();
-	texture[CameraServer::FEED_Y_IMAGE] = vs->texture_create(); // also used for RGBA
-	texture[CameraServer::FEED_CBCR_IMAGE] = vs->texture_create();
-#endif
 }
 
 CameraFeed::~CameraFeed() {
-// FIXME: Disabled during Vulkan refactoring, should be ported.
-#if 0
 	// Free our textures
-	RenderingServer *vs = RenderingServer::get_singleton();
-	vs->free(texture[CameraServer::FEED_Y_IMAGE]);
-	vs->free(texture[CameraServer::FEED_CBCR_IMAGE]);
-#endif
+	if (texture[CameraServer::FEED_Y_IMAGE].is_null()) {
+		RenderingServer::get_singleton()->free(texture[CameraServer::FEED_Y_IMAGE]);
+	}
+	if (texture[CameraServer::FEED_CBCR_IMAGE].is_null()) {
+		RenderingServer::get_singleton()->free(texture[CameraServer::FEED_CBCR_IMAGE]);
+	}
 }
 
 void CameraFeed::set_RGB_img(const Ref<Image> &p_rgb_img) {
-// FIXME: Disabled during Vulkan refactoring, should be ported.
-#if 0
 	ERR_FAIL_COND(p_rgb_img.is_null());
 	if (active) {
-		RenderingServer *vs = RenderingServer::get_singleton();
-
 		int new_width = p_rgb_img->get_width();
 		int new_height = p_rgb_img->get_height();
 
@@ -198,23 +174,22 @@ void CameraFeed::set_RGB_img(const Ref<Image> &p_rgb_img) {
 			// We're assuming here that our camera image doesn't change around formats etc, allocate the whole lot...
 			base_width = new_width;
 			base_height = new_height;
+		}
 
-			vs->texture_allocate(texture[CameraServer::FEED_RGBA_IMAGE], new_width, new_height, 0, Image::FORMAT_RGB8, RS::TEXTURE_TYPE_2D, RS::TEXTURE_FLAGS_DEFAULT);
+		if (texture[CameraServer::FEED_RGBA_IMAGE].is_null()) {
+			texture[CameraServer::FEED_RGBA_IMAGE] = RenderingServer::get_singleton()->texture_2d_create(p_rgb_img);
+		} else {
+			RID new_texture = RenderingServer::get_singleton()->texture_2d_create(p_rgb_img);
+			RenderingServer::get_singleton()->texture_replace(texture[CameraServer::FEED_RGBA_IMAGE], new_texture);
 		}
 
-		vs->texture_set_data(texture[CameraServer::FEED_RGBA_IMAGE], p_rgb_img);
 		datatype = CameraFeed::FEED_RGB;
 	}
-#endif
 }
 
 void CameraFeed::set_YCbCr_img(const Ref<Image> &p_ycbcr_img) {
-// FIXME: Disabled during Vulkan refactoring, should be ported.
-#if 0
 	ERR_FAIL_COND(p_ycbcr_img.is_null());
 	if (active) {
-		RenderingServer *vs = RenderingServer::get_singleton();
-
 		int new_width = p_ycbcr_img->get_width();
 		int new_height = p_ycbcr_img->get_height();
 
@@ -222,67 +197,53 @@ void CameraFeed::set_YCbCr_img(const Ref<Image> &p_ycbcr_img) {
 			// We're assuming here that our camera image doesn't change around formats etc, allocate the whole lot...
 			base_width = new_width;
 			base_height = new_height;
+		}
 
-			vs->texture_allocate(texture[CameraServer::FEED_RGBA_IMAGE], new_width, new_height, 0, Image::FORMAT_RGB8, RS::TEXTURE_TYPE_2D, RS::TEXTURE_FLAGS_DEFAULT);
+		if (texture[CameraServer::FEED_RGBA_IMAGE].is_null()) {
+			texture[CameraServer::FEED_RGBA_IMAGE] = RenderingServer::get_singleton()->texture_2d_create(p_ycbcr_img);
+		} else {
+			RID new_texture = RenderingServer::get_singleton()->texture_2d_create(p_ycbcr_img);
+			RenderingServer::get_singleton()->texture_replace(texture[CameraServer::FEED_RGBA_IMAGE], new_texture);
 		}
 
-		vs->texture_set_data(texture[CameraServer::FEED_RGBA_IMAGE], p_ycbcr_img);
 		datatype = CameraFeed::FEED_YCBCR;
 	}
-#endif
 }
 
 void CameraFeed::set_YCbCr_imgs(const Ref<Image> &p_y_img, const Ref<Image> &p_cbcr_img) {
-// FIXME: Disabled during Vulkan refactoring, should be ported.
-#if 0
 	ERR_FAIL_COND(p_y_img.is_null());
 	ERR_FAIL_COND(p_cbcr_img.is_null());
 	if (active) {
-		RenderingServer *vs = RenderingServer::get_singleton();
-
 		///@TODO investigate whether we can use thirdparty/misc/yuv2rgb.h here to convert our YUV data to RGB, our shader approach is potentially faster though..
 		// Wondering about including that into multiple projects, may cause issues.
 		// That said, if we convert to RGB, we could enable using texture resources again...
 
 		int new_y_width = p_y_img->get_width();
 		int new_y_height = p_y_img->get_height();
-		int new_cbcr_width = p_cbcr_img->get_width();
-		int new_cbcr_height = p_cbcr_img->get_height();
 
 		if ((base_width != new_y_width) || (base_height != new_y_height)) {
 			// We're assuming here that our camera image doesn't change around formats etc, allocate the whole lot...
 			base_width = new_y_width;
 			base_height = new_y_height;
+		}
 
-			vs->texture_allocate(texture[CameraServer::FEED_Y_IMAGE], new_y_width, new_y_height, 0, Image::FORMAT_R8, RS::TEXTURE_TYPE_2D, RS::TEXTURE_FLAG_USED_FOR_STREAMING);
+		if (texture[CameraServer::FEED_Y_IMAGE].is_null()) {
+			texture[CameraServer::FEED_Y_IMAGE] = RenderingServer::get_singleton()->texture_2d_create(p_y_img);
+		} else {
+			RID new_texture = RenderingServer::get_singleton()->texture_2d_create(p_y_img);
+			RenderingServer::get_singleton()->texture_replace(texture[CameraServer::FEED_Y_IMAGE], new_texture);
+		}
 
-			///@TODO GLES2 doesn't support FORMAT_RG8, need to do some form of conversion
-			vs->texture_allocate(texture[CameraServer::FEED_CBCR_IMAGE], new_cbcr_width, new_cbcr_height, 0, Image::FORMAT_RG8, RS::TEXTURE_TYPE_2D, RS::TEXTURE_FLAG_USED_FOR_STREAMING);
+		if (texture[CameraServer::FEED_CBCR_IMAGE].is_null()) {
+			texture[CameraServer::FEED_CBCR_IMAGE] = RenderingServer::get_singleton()->texture_2d_create(p_cbcr_img);
+		} else {
+			RID new_texture = RenderingServer::get_singleton()->texture_2d_create(p_cbcr_img);
+			RenderingServer::get_singleton()->texture_replace(texture[CameraServer::FEED_CBCR_IMAGE], new_texture);
 		}
 
-		vs->texture_set_data(texture[CameraServer::FEED_Y_IMAGE], p_y_img);
-		vs->texture_set_data(texture[CameraServer::FEED_CBCR_IMAGE], p_cbcr_img);
 		datatype = CameraFeed::FEED_YCBCR_SEP;
 	}
-#endif
-}
-
-// FIXME: Disabled during Vulkan refactoring, should be ported.
-#if 0
-void CameraFeed::allocate_texture(int p_width, int p_height, Image::Format p_format, RenderingServer::TextureType p_texture_type, FeedDataType p_data_type) {
-	RenderingServer *vs = RenderingServer::get_singleton();
-
-	if ((base_width != p_width) || (base_height != p_height)) {
-		// We're assuming here that our camera image doesn't change around formats etc, allocate the whole lot...
-		base_width = p_width;
-		base_height = p_height;
-
-		vs->texture_allocate(texture[0], p_width, p_height, 0, p_format, p_texture_type, RS::TEXTURE_FLAGS_DEFAULT);
-	}
-
-	datatype = p_data_type;
 }
-#endif
 
 bool CameraFeed::activate_feed() {
 	// nothing to do here
diff --git a/servers/camera/camera_feed.h b/servers/camera/camera_feed.h
index fac7ae1b2d..1f80d15e30 100644
--- a/servers/camera/camera_feed.h
+++ b/servers/camera/camera_feed.h
@@ -103,10 +103,6 @@ public:
 	void set_RGB_img(const Ref<Image> &p_rgb_img);
 	void set_YCbCr_img(const Ref<Image> &p_ycbcr_img);
 	void set_YCbCr_imgs(const Ref<Image> &p_y_img, const Ref<Image> &p_cbcr_img);
-// FIXME: Disabled during Vulkan refactoring, should be ported.
-#if 0
-	void allocate_texture(int p_width, int p_height, Image::Format p_format, RenderingServer::TextureType p_texture_type, FeedDataType p_data_type);
-#endif
 
 	virtual bool activate_feed();
 	virtual void deactivate_feed();
```

**Comment by bruvzg** [comment]
PR for CameraFeed fix - #51581

**Comment by RMKD** [comment]
Confirmed to work via the lib4vl2 method on linux in-editor (thx for the demo scene in the PR) - but the texture doesn't render in a played scene, even with a script to cycle through and `set_active(true)` so I'll see if I can sort out what's going on there.

**Comment by bruvzg** [comment]
> Confirmed to work via the lib4vl2 method on linux in-editor (thx for the demo scene in the PR) - but the texture doesn't render in a played scene, even with a script to cycle through and set_active(true) so I'll see if I can sort out what's going on there.

It seems to work only 3D scene is opened, closed and reopened in editor, or if it's assigned to 2D sprite (in this case it's working in running project, but only after resizing the window).

Probably something in the CameraTerxture/Feed/Server is initialized before RenderingServer is ready, but I'm not sure what's exactly wrong.

**Comment by pkowal1982** [comment]
I was also implementing linux camera support, so maybe we could combine best of both solutions?
https://github.com/pkowal1982/godot/tree/camera

**Comment by CrispyPin** [comment]
What testing is needed for this to be merged?

What is required for this to be backported to 3.x?

**Comment by Schmetzler** [comment]
@CrispyPin 

See my note above:
> Ah as I want to mention. I would also like to push it to the godot 3.x branch, but therefore the Vector<uint8_t> must be changed to PoolVector<uint8_t>. But i guess it should be cherrypicked after this one is merged?

And according to this you have to change `void V4l2_Device::get_image(Ref<CameraFeed> feed, uint8_t *buffer)`  in `camera_x11.cpp` around line 572. 

from:
```
uint8_t *w = img_data.ptrw();
// TODO: Buffer is 1024 Byte longer?
memcpy(w, buffer, width * height * 3);
```

to:

```
PoolVector<uint8_t>::Write w = img_data.write();
// TODO: Buffer is 1024 Byte longer?
memcpy(w.ptr(), buffer, width*height*3);
```

This should be enough to backport it to godot 3.x

**Comment by bnolan** [comment]
Will this be merged into 3.5?

**Comment by akien-mga** [comment]
Superseded by #53666. Thanks for the contribution nonetheless!

**Pull request closed without merging**